make local-install error

(1) Issue/Bug Description:

make local-install throws the following error:
src/extension.ts:1007:25 - error TS6133: '_entity' is declared but its value is never read.

1007             for (const [_entity, compare] of this.windows.iter()) {
                             ~~~~~~~


Found 1 error.

make: *** [Makefile:38: compile] Error 2


(2) Steps to reproduce (if you know):

Following the instructions for Using Pop Shell on other GNOME Desktops for Ubuntu-Based Distros

(3) Expected behavior:

Pop Shell's installed

(4) Distribution (run cat /etc/os-release):

NAME="Ubuntu"
VERSION="20.04.4 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.4 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal

(7) Where was Pop Shell installed from:

https://github.com/pop-os/shell

(8) Monitor Setup (2 x 1080p, 4K, Primary(Horizontal), Secondary(Vertical), etc):

(9) Other Installed/Enabled Extensions:

(10) Other Notes:

--- a/src/extension.ts
+++ b/src/extension.ts
@@ -741,6 +741,11 @@ export class Ext extends Ecs.System<ExtEvent> {
         return Rect.Rectangle.from_meta(meta as Rectangular);
     }
 
+    monitor_area(monitor: number): Rectangle | null {
+        const rect = global.display.get_monitor_geometry(monitor)
+        return rect ? Rect.Rectangle.from_meta(rect as Rectangular) : null
+    }
+
     on_active_workspace_changed() {
         this.register_fn(() => {
             this.exit_modes();

--- a/src/extension.ts
+++ b/src/extension.ts
@@ -752,6 +752,7 @@ export class Ext extends Ecs.System<ExtEvent> {
             this.restack()
 
             const activate_window = (window: Window.ShellWindow) => {
+                this.on_focused(window)
                 window.activate(true)
                 this.prev_focused = [null, window.entity]
             }
@@ -761,6 +762,7 @@ export class Ext extends Ecs.System<ExtEvent> {
                 activate_window(focused)
                 return
             }
+
 
             // Activate the last-active window on workspace.
             const workspace_id = this.active_workspace()

--- a/src/extension.ts
+++ b/src/extension.ts
@@ -895,13 +895,8 @@ export class Ext extends Ecs.System<ExtEvent> {
                     if (prev.is_maximized()) {
                         prev.meta.unmaximize(Meta.MaximizeFlags.BOTH);
                     }
-
-                    if (prev.meta.is_fullscreen()) {
-                        prev.meta.unmake_fullscreen();
-                    }
                 } else if (prev.stack) {
                     prev.meta.unmaximize(Meta.MaximizeFlags.BOTH)
-                    prev.meta.unmake_fullscreen()
                     this.auto_tiler.forest.stacks.get(prev.stack)?.restack()
                 }
             }
@@ -1018,10 +1013,6 @@ export class Ext extends Ecs.System<ExtEvent> {
                 if (is_same_space && !this.contains_tag(compare.entity, Tags.Floating)) {
                     if (compare.is_maximized()) {
                         compare.meta.unmaximize(Meta.MaximizeFlags.BOTH);
-                    }
-
-                    if (compare.meta.is_fullscreen()) {
-                        compare.meta.unmake_fullscreen();
                     }
                 }
             }
@@ -1266,8 +1257,6 @@ export class Ext extends Ecs.System<ExtEvent> {
     move_monitor(direction: Meta.DisplayDirection) {
         const win = this.focus_window();
         if (!win) return;
-        if (win && win.meta.is_fullscreen())
-            win.meta.unmake_fullscreen();
 
         const prev_monitor = win.meta.get_monitor();
         const next_monitor = Tiling.locate_monitor(win, direction);
@@ -1289,8 +1278,6 @@ export class Ext extends Ecs.System<ExtEvent> {
     move_workspace(direction: Meta.DisplayDirection) {
         const win = this.focus_window();
         if (!win) return;
-        if (win && win.meta.is_fullscreen())
-            win.meta.unmake_fullscreen();
 
         /** Move a window between workspaces */
         const workspace_move = (direction: Meta.MotionDirection) => {
@@ -1546,13 +1533,6 @@ export class Ext extends Ecs.System<ExtEvent> {
             // Raise maximized to top so stacks won't appear over them.
             const actor = win.meta.get_compositor_private();
             if (actor) global.window_group.set_child_above_sibling(actor, null);
-
-            if (win.meta.is_fullscreen()) {
-                this.size_changed_block();
-                win.meta.unmake_fullscreen();
-                win.meta.maximize(Meta.MaximizeFlags.BOTH);
-                this.size_changed_unblock();
-            }
 
             this.on_monitor_changed(win, (_cfrom, cto, workspace) => {
                 if (win) {

--- a/src/extension.ts
+++ b/src/extension.ts
@@ -5,6 +5,7 @@ import * as Forest from 'forest';
 import * as Ecs from 'ecs';
 import * as Events from 'events';
 import * as Focus from 'focus';
+import * as Geom from 'geom';
 import * as GrabOp from 'grab_op';
 import * as Keybindings from 'keybindings';
 import * as Lib from 'lib';
@@ -28,7 +29,7 @@ import * as scheduler from 'scheduler';
 
 import type { Entity } from 'ecs';
 import type { ExtEvent } from 'events';
-import type { Rectangle } from 'rectangle';
+import { Rectangle } from 'rectangle';
 import type { Indicator } from 'panel_settings';
 import type { Launcher } from 'launcher';
 
@@ -1250,6 +1251,7 @@ export class Ext extends Ecs.System<ExtEvent> {
                     win.meta.unmaximize(Meta.MaximizeFlags.VERTICAL);
                     win.meta.unmaximize(Meta.MaximizeFlags.BOTH);
                 }
+
                 this.register(Events.window_move(this, win, rect));
             } else {
                 win.move(this, rect, () => {});
@@ -1300,12 +1302,45 @@ export class Ext extends Ecs.System<ExtEvent> {
                 return last;
             }
 
+            /** Places window onto the nearest window of a given workspace */
+            const place_on_nearest_window = (auto_tiler: auto_tiler.AutoTiler, ws: Meta.Workspace, monitor: number) => {
+                const src = win.meta.get_frame_rect()
+
+                auto_tiler.detach_window(this, win.entity);
+
+                const index = ws.index()
+                const coord: [number, number] = [src.x, src.y]
+                
+                let nearest_window = null
+                let nearest_distance = null
+
+                for (const [entity, window] of this.windows.iter()) {
+                    const other_monitor = window.meta.get_monitor()
+                    const other_index = window.meta.get_workspace().index()
+                    if (!this.contains_tag(entity, Tags.Floating) && other_monitor == monitor && other_index === index && !Ecs.entity_eq(win.entity, window.entity)) {
+                        const other_rect = window.rect()
+                        const other_coord: [number, number] = [other_rect.x, other_rect.y]
+                        const distance = Geom.distance(coord, other_coord)
+                        if (nearest_distance === null || nearest_distance > distance) {
+                            nearest_window = window
+                            nearest_distance = distance
+                        }
+                    }
+                }
+
+                if (nearest_window === null) {
+                    auto_tiler.attach_to_workspace(this, win, [monitor, index]);
+                } else {
+                    auto_tiler.attach_to_window(this, nearest_window, win, { src }, false)
+                }
+            }
+
             const move_to_neighbor = (neighbor: Meta.Workspace) => {
                 const monitor = win.meta.get_monitor();
                 if (this.auto_tiler && win.is_tilable(this)) {
                     win.ignore_detach = true;
-                    this.auto_tiler.detach_window(this, win.entity);
-                    this.auto_tiler.attach_to_workspace(this, win, [monitor, neighbor.index()]);
+                    
+                    place_on_nearest_window(this.auto_tiler, neighbor, monitor)
 
                     if (win.meta.minimized) {
                         this.size_signals_block(win);

--- a/src/extension.ts
+++ b/src/extension.ts
@@ -1007,7 +1007,7 @@ export class Ext extends Ecs.System<ExtEvent> {
             for (const [_entity, compare] of this.windows.iter()) {
                 const is_same_space = compare.meta.get_monitor() === mon
                     && compare.meta.get_workspace().index() === work;
-                if (is_same_space) {
+                if (is_same_space && !this.contains_tag(compare.entity, Tags.Floating)) {
                     if (compare.is_maximized()) {
                         compare.meta.unmaximize(Meta.MaximizeFlags.BOTH);
                     }

--- a/src/extension.ts
+++ b/src/extension.ts
@@ -1012,7 +1012,7 @@ export class Ext extends Ecs.System<ExtEvent> {
                 work = win.meta.get_workspace().index()
             }
 
-            for (const [_entity, compare] of this.windows.iter()) {
+            for (const [, compare] of this.windows.iter()) {
                 const is_same_space = compare.meta.get_monitor() === mon
                     && compare.meta.get_workspace().index() === work;
                 if (is_same_space && !this.contains_tag(compare.entity, Tags.Floating)) {
@@ -1310,7 +1310,7 @@ export class Ext extends Ecs.System<ExtEvent> {
 
                 const index = ws.index()
                 const coord: [number, number] = [src.x, src.y]
-                
+
                 let nearest_window = null
                 let nearest_distance = null
 
@@ -1339,7 +1339,7 @@ export class Ext extends Ecs.System<ExtEvent> {
                 const monitor = win.meta.get_monitor();
                 if (this.auto_tiler && win.is_tilable(this)) {
                     win.ignore_detach = true;
-                    
+
                     place_on_nearest_window(this.auto_tiler, neighbor, monitor)
 
                     if (win.meta.minimized) {

--- a/src/forest.ts
+++ b/src/forest.ts
@@ -264,6 +264,12 @@ export class Forest extends Ecs.World {
                 } else {
                     fork.right = right_node;
                     fork.set_ratio(fork.length() / 2);
+
+                    if ("src" in place_by) {
+                        const [left, right] = area_of_halves(fork)
+                        place_by_keyboard(fork, place_by.src, left, right)
+                    }
+
                     return this._attach(onto_entity, new_entity, this.on_attach, entity, fork, null);
                 }
             } else if (fork.left.is_in_stack(onto_entity)) {

--- a/src/keybindings.ts
+++ b/src/keybindings.ts
@@ -16,6 +16,11 @@ export class Keybindings {
         this.ext = ext;
         this.global = {
             "activate-launcher": () => {
+                // Ignore when a fullscreen window is active
+                if (ext.focus_window()?.meta.is_fullscreen()) {
+                    return
+                }
+
                 ext.tiler.exit(ext);
                 ext.window_search.open(ext);
             },

--- a/src/launcher.ts
+++ b/src/launcher.ts
@@ -263,7 +263,7 @@ export class Launcher extends search.Search {
     }
 
     load_desktop_files() {
-        log.warn("pop-shell: deprecated function called (dialog_launcher::load_desktop_files)")
+        log.warn("pop-shell: deprecated function called (launcher::load_desktop_files)")
     }
 
     locate_by_app_info(info: any): null | ShellWindow {
@@ -293,10 +293,16 @@ export class Launcher extends search.Search {
     }
 
     open(ext: Ext) {
+        // Do not activate if the focused window is fullscreen
+        if (ext.focus_window()?.meta.is_fullscreen()) {
+            return
+        }
+
         const active_monitor = ext.active_monitor()
         const mon_work_area = ext.monitor_work_area(active_monitor)
         const mon_area = ext.monitor_area(active_monitor)
         const mon_width = mon_area ? mon_area.width : mon_work_area.width
+
 
         super.cleanup()
 

--- a/src/launcher.ts
+++ b/src/launcher.ts
@@ -293,7 +293,10 @@ export class Launcher extends search.Search {
     }
 
     open(ext: Ext) {
-        const mon = ext.monitor_work_area(ext.active_monitor())
+        const active_monitor = ext.active_monitor()
+        const mon_work_area = ext.monitor_work_area(active_monitor)
+        const mon_area = ext.monitor_area(active_monitor)
+        const mon_width = mon_area ? mon_area.width : mon_work_area.width
 
         super.cleanup()
 
@@ -302,9 +305,9 @@ export class Launcher extends search.Search {
 
         super._open(global.get_current_time(), false)
 
-        this.dialog.dialogLayout.x = (mon.width / 2) - (this.dialog.dialogLayout.width / 2)
+        this.dialog.dialogLayout.x = (mon_width / 2) - (this.dialog.dialogLayout.width / 2)
 
-        let height = mon.height >= 900 ? mon.height / 2 : mon.height / 3.5
+        let height = mon_work_area.height >= 900 ? mon_work_area.height / 2 : mon_work_area.height / 3.5
         this.dialog.dialogLayout.y = height - (this.dialog.dialogLayout.height / 2)
     }
 

--- a/src/mod.d.ts
+++ b/src/mod.d.ts
@@ -220,7 +220,6 @@ declare namespace Meta {
         move_resize_frame(user_op: boolean, x: number, y: number, w: number, h: number): boolean;
         raise(): void;
         skip_taskbar: boolean;
-        unmake_fullscreen(): void;
         unmaximize(flags: any): void;
         unminimize(): void;
     }

--- a/src/tiling.ts
+++ b/src/tiling.ts
@@ -201,10 +201,6 @@ export class Tiler {
             this.queue.send(() => {
                 const focused = ext.focus_window();
                 if (focused) {
-                    if (focused.meta.is_fullscreen()) {
-                        focused.meta.unmake_fullscreen();
-                    }
-
                     // The window that the focused window is being moved onto
                     const move_to = focus();
 

--- a/src/window.ts
+++ b/src/window.ts
@@ -417,7 +417,8 @@ export class ShellWindow {
             let border = this.border;
 
             const permitted = () => {
-                return this.ext.focus_window() == this
+                return this.actor_exists()
+                    && this.ext.focus_window() == this
                     && !this.meta.is_fullscreen()
                     && (!this.is_single_max_screen() || this.is_snap_edge())
                     && !this.meta.minimized
@@ -427,16 +428,19 @@ export class ShellWindow {
                 if (this.meta.appears_focused) {
                     border.show();
 
+                    // Focus will be re-applied to fix windows moving across workspaces
+                    let applications = 0
+
                     // Ensure that the border is shown
                     if (ACTIVE_HINT_SHOW_ID !== null) GLib.source_remove(ACTIVE_HINT_SHOW_ID)
                     ACTIVE_HINT_SHOW_ID = GLib.timeout_add(GLib.PRIORITY_DEFAULT, 600, () => {
-                        if (!this.same_workspace() || border.visible || !permitted()) {
+                        if (applications > 4 && !this.same_workspace() || !permitted()) {
                             ACTIVE_HINT_SHOW_ID = null
                             return GLib.SOURCE_REMOVE
                         }
 
+                        applications += 1
                         border.show()
-
                         return GLib.SOURCE_CONTINUE
                     })
                 }

--- a/src/window.ts
+++ b/src/window.ts
@@ -659,10 +659,21 @@ export function activate(ext: Ext, move_mouse: boolean, default_pointer_position
         && imports.ui.main.modalCount === 0
         && ext.settings.mouse_cursor_follows_active_window()
         && !pointer_already_on_window(win)
+        && pointer_in_work_area()
 
     if (pointer_placement_permitted) {
         place_pointer_on(default_pointer_position, win)
     }
+}
+
+function pointer_in_work_area(): boolean {
+    const cursor = lib.cursor_rect()
+    const indice = global.display.get_current_monitor()
+    const mon = global.display.get_workspace_manager()
+        .get_active_workspace()
+        .get_work_area_for_monitor(indice)
+
+    return mon ? cursor.intersects(mon) : false
 }
 
 export function place_pointer_on(default_pointer_position: Config.DefaultPointerPosition, win: Meta.Window) {

--- a/src/window.ts
+++ b/src/window.ts
@@ -247,10 +247,8 @@ export class ShellWindow {
     }
 
     is_maximized(): boolean {
-        let maximized = this.meta.get_maximized() !== 0;
-        return maximized;
+        return this.meta.get_maximized() !== 0
     }
-
 
     /**
      * Window is maximized, 0 gapped or smart gapped
@@ -419,10 +417,10 @@ export class ShellWindow {
             let border = this.border;
 
             const permitted = () => {
-                return !this.meta.is_fullscreen() &&
-                    (!this.is_single_max_screen() || this.is_snap_edge()) &&
-                    !this.meta.minimized &&
-                    this.same_workspace()
+                return this.ext.focus_window() == this
+                    && !this.meta.is_fullscreen()
+                    && (!this.is_single_max_screen() || this.is_snap_edge())
+                    && !this.meta.minimized
             }
 
             if (permitted()) {

--- a/src/window.ts
+++ b/src/window.ts
@@ -94,6 +94,11 @@ export class ShellWindow {
 
         this.known_workspace = this.workspace_id()
 
+        // Float fullscreen windows by default, such as Kodi.
+        if (this.meta.is_fullscreen()) {
+            ext.add_tag(entity, Tags.Floating)
+        }
+
         if (this.may_decorate()) {
             if (!window.is_client_decorated()) {
                 if (ext.settings.show_title()) {
@@ -308,7 +313,7 @@ export class ShellWindow {
                 // Transient windows are most likely dialogs
                 && !this.is_transient()
                 // If a window lacks a class, it's probably a web browser dialog
-                && wm_class !== null;
+                && wm_class !== null
         };
 
         return !ext.contains_tag(this.entity, Tags.Floating)

